src/clients/UBAClient/UBAClientBase.ts

```diff
@@ -0,0 +1,346 @@
+   */
+  public getFlows(chainId: number, tokenSymbol: string, fromBlock?: number, toBlock?: number): UbaFlow[] {
+    const relevantBundleStates = this.retrieveBundleStates(chainId, tokenSymbol);
+    return relevantBundleStates
+      .flatMap((bundleState) => bundleState.flows)
+      .map((flow) => flow.flow)
+      .filter(
+        (flow) =>
+          (fromBlock === undefined || flow.blockNumber >= fromBlock) &&
+          (toBlock === undefined || flow.blockNumber <= toBlock)
+      );
+  }
+
+  /**
+   * @description Evaluate an RefundRequest object for validity.
+   * @dev  Callers should evaluate 'valid' before 'reason' in the return object.
+   * @dev  The following RefundRequest attributes are not evaluated for validity and should be checked separately:
+   * @dev  - previousIdenticalRequests
+   * @dev  - Age of blockNumber (i.e. according to SpokePool finality)
+   * @param chainId       ChainId of SpokePool where refundRequest originated.
+   * @param refundRequest RefundRequest object to be evaluated for validity.
+   */
+  public refundRequestIsValid(chainId: number, refundRequest: RefundRequestWithBlock): RequestValidReturnType {
+    const result = this.getFlows(chainId, refundRequest.refundToken).some((flow) => {
+      return (
+        flow.logIndex == refundRequest.logIndex &&
+        flow.blockNumber == refundRequest.blockNumber &&
+        flow.transactionHash == refundRequest.transactionHash &&
+        flow.transactionIndex == refundRequest.transactionIndex
+      );
+    });
+    if (!result) {
+      return {
+        valid: false,
+        reason: "RefundRequest is not a valid flow because it didn't appear in the list of validated flows.",
+      };
+    } else {
+      return {
+        valid: true,
+      };
+    }
+  }
+
+  /**
+   * Calculate the balancing fee of a given token on a given chainId at a given block number
+   * @param tokenSymbol The token to get the balancing fee for
+   * @param amount The amount to get the balancing fee for
+   * @param balancingActionBlockNumber The block number to get the balancing fee for
+   * @param chainId The chainId to get the balancing fee for. If the feeType is Deposit, this is the deposit chainId. If the feeType is Refund, this is the refund chainId.
+   * @param feeType The type of fee to calculate
+   * @returns The balancing fee for the given token on the given chainId at the given block number
+   */
+  public computeBalancingFee(
+    tokenSymbol: string,
+    amount: BigNumber,
+    balancingActionBlockNumber: number,
+    chainId: number,
+    feeType: UBAActionType
+  ): BalancingFeeReturnType {
+    // Opening balance for the balancing action blockNumber.
+    const relevantBundleStates = this.retrieveBundleStates(chainId, tokenSymbol);
+    const specificBundleState = relevantBundleStates.findLast(
+      (bundleState) => bundleState.blockNumber <= balancingActionBlockNumber
```

Comment:
I keep getting a bit confused about which `chainId` the `bundleState.blockNumber` is referring to: is it the Ethereum block at which the proposal happened, or the `bundleEndBlock` for the `chainId` that was proposed in the bundle? I assume the latter, so maybe can we rename the param to `endBlockForChain`?

Reply:
I think this is contained in: https://github.com/across-protocol/sdk-v2/pull/266#discussion_r1248156998